### PR TITLE
feat: group member expiration

### PR DESCRIPTION
**Pull request by j75689**
### Description

This PR is to introduce to the GroupMember Expiration feature.

### Rationale

tell us why we need these changes...

### Example

add an example CLI or API response...

### Changes

Notable changes: 
* storage module

### CODE REVIEW

File: x/storage/types/crosschain.go [hunk 1]
```diff
@@ -856,6 +859,7 @@ var (
 		{Name: "OperationType", Type: "uint8"},
 		{Name: "Members", Type: "address[]"},
 		{Name: "ExtraData", Type: "bytes"},
+		{Name: "MemberExpiration", Type: "uint64[]"},
```
Comment (guagualvcha) [inline x/storage/types/crosschain.go:862]: Is it compatible to add this when do abi decode?

File: x/storage/keeper/msg_server.go [hunk 2]
```diff
@@ -232,7 +232,7 @@ func (k msgServer) CreateGroup(goCtx context.Context, msg *types.MsgCreateGroup)
 
 	ownerAcc := sdk.MustAccAddressFromHex(msg.Creator)
 
-	id, err := k.Keeper.CreateGroup(ctx, ownerAcc, msg.GroupName, CreateGroupOptions{Members: msg.Members, Extra: msg.Extra})
+	id, err := k.Keeper.CreateGroup(ctx, ownerAcc, msg.GroupName, CreateGroupOptions{Members: msg.Members, Extra: msg.Extra, MemberWithExpirationTime: msg.MembersWithExpiration})
```
[the remark this answers is not in the document]
Comment (guagualvcha) [reply x/storage/keeper/msg_server.go:235]: So is the tx `MsgUpdateGroupMember`

**Comment by j75689** [comment]
Offline cmd compatible New/old version msg needs to be discussed

File: x/permission/keeper/keeper.go [hunk 3]
```diff
@@ -71,9 +72,37 @@ func (k Keeper) AddGroupMember(ctx sdk.Context, groupID math.Uint, member sdk.Ac
...
+		return err
+	}
+
+	store := ctx.KVStore(k.storeKey)
+	// We can simply override the whole value here, because the expiration time is the only field in the value.
+	// If there are more fields in the future, we should use a more sophisticated way to update the value.
+	memberExtra := types.GroupMemberExtra{
+		ExpirationTime: expiration,
+	}
+	store.Set(types.GetGroupMemberExtraKey(groupID, member), id.Bytes())
```
Comment (guagualvcha) [inline x/permission/keeper/keeper.go:90]: Why we have to set two storage here, is it possible to reduce to one?

File: x/permission/keeper/keeper.go [hunk 4]
```diff
@@ -71,9 +72,37 @@ func (k Keeper) AddGroupMember(ctx sdk.Context, groupID math.Uint, member sdk.Ac
...
 }
 
+func (k Keeper) UpdateGroupMemberExpiration(ctx sdk.Context, groupID math.Uint, member sdk.AccAddress, memberID math.Uint, expiration time.Time) {
+	store := ctx.KVStore(k.storeKey)
+	// We can simply override the whole value here, because the expiration time is the only field in the value.
+	// If there are more fields in the future, we should use a more sophisticated way to update the value.
+	memberExtra := types.GroupMemberExtra{
+		ExpirationTime: expiration,
+	}
+	store.Set(types.GetGroupMemberExtraKey(groupID, member), memberID.Bytes())
```
Comment (guagualvcha) [inline x/permission/keeper/keeper.go:102]: If only update, is this necessary?

File: x/storage/keeper/cross_app_group.go [hunk 5]
```diff
@@ -374,6 +377,94 @@ func (app *GroupApp) handleUpdateGroupMemberSynPackage(ctx sdk.Context, header *
...
+				Operator:  updateGroupPackage.Operator,
+				ExtraData: updateGroupPackage.ExtraData,
+			}.MustSerialize(),
+			Err: types.ErrNoSuchGroup,
+		}
+	}
+
+	switch updateGroupPackage.OperationType {
+	case types.OperationAddGroupMember, types.OperationDeleteGroupMember:
+		err = app.handleAddOrDeleteGroupMemberOperation(ctx, groupInfo, updateGroupPackage)
```
Comment (guagualvcha) [inline x/storage/keeper/cross_app_group.go:407]: if it is add member, it is possible to have expire time, but can not find it here now.

File: x/permission/keeper/keeper.go (hunk 3 above)
Comment (j75689) [reply x/permission/keeper/keeper.go]: yes, merge it to `GroupMember`

File: x/permission/keeper/keeper.go (hunk 4 above)
Comment (j75689) [reply x/permission/keeper/keeper.go]: fixed

File: x/storage/keeper/cross_app_group.go (hunk 5 above)
Comment (j75689) [reply x/storage/keeper/cross_app_group.go:407]: resolved, allow creation group Member with expiration from cross-chain package now

File: proto/greenfield/storage/events.proto [hunk 6]
```diff
@@ -338,6 +341,41 @@ message EventUpdateGroupMember {
   repeated string members_to_add = 5 [(cosmos_proto.scalar) = "cosmos.AddressString"];
   // members_to_add defines all the members to be deleted from the group
   repeated string members_to_delete = 6 [(cosmos_proto.scalar) = "cosmos.AddressString"];
+  // added_members_detail defines the all the members detail of the group.
+  repeated EventGroupMemberDetail added_members_detail = 7;
```
Comment (guagualvcha) [inline proto/greenfield/storage/events.proto:345]: if we get this, then `members_to_add` is not necessary.

File: proto/greenfield/storage/events.proto [hunk 7]
```diff
@@ -338,6 +341,41 @@ message EventUpdateGroupMember {
...
+  // id define an u256 id for group
+  string group_id = 4 [
+    (cosmos_proto.scalar) = "cosmos.Uint",
+    (gogoproto.customtype) = "Uint",
+    (gogoproto.nullable) = false
+  ];
+  // source_type define the source of the group. CrossChain or Greenfield origin
+  SourceType source_type = 5;
+  // members define the all the address of the members.
+  repeated string members = 6 [(cosmos_proto.scalar) = "cosmos.AddressString"];
```
Comment (guagualvcha) [inline proto/greenfield/storage/events.proto:364]: No need for this field.

File: x/permission/keeper/keeper.go [hunk 8]
```diff
@@ -283,14 +296,21 @@ func (k Keeper) VerifyPolicy(ctx sdk.Context, resourceID math.Uint, resourceType
...
+				groupMember, memberFound := k.GetGroupMember(ctx, item.GroupId, operator)
 				if memberFound {
 					if effect == types.EFFECT_ALLOW {
 						allowed = true
 					} else if effect == types.EFFECT_DENY {
 						return types.EFFECT_DENY
 					}
 				}
+
+				// post check if the operator has not been expired
```
Comment (guagualvcha) [inline x/permission/keeper/keeper.go:308]: is it possble that `ExpirationTime` is not nil, but with empty vaule?

File: x/storage/keeper/keeper.go [hunk 9]
```diff
@@ -1254,24 +1261,32 @@ func (k Keeper) CreateGroup(
...
 
+	addedMembersDetailEvent := make([]*types.EventGroupMemberDetail, 0, len(opts.Members))
 	// need to limit the size of Msg.Members to avoid taking too long to execute the msg
 	for _, member := range opts.Members {
 		memberAddress, err := sdk.AccAddressFromHexUnsafe(member)
 		if err != nil {
 			return sdkmath.ZeroUint(), err
 		}
-		err = k.permKeeper.AddGroupMember(ctx, groupInfo.Id, memberAddress)
+		err = k.permKeeper.AddGroupMember(ctx, groupInfo.Id, memberAddress, nil)
```
Comment (guagualvcha) [inline x/storage/keeper/keeper.go:1271]: A bit strange that we support create a group with initial `members` but without expire time, I suggest that we either dont support any member when create group or we should support expire time of member.

I prefer the first option

File: x/storage/keeper/keeper.go [hunk 10]
```diff
@@ -1409,20 +1440,74 @@ func (k Keeper) UpdateGroupMember(ctx sdk.Context, operator sdk.AccAddress, grou
...
+	}
+
+	if err := ctx.EventManager().EmitTypedEvents(&types.EventRenewGroupMember{
+		Operator:      operator.String(),
+		Owner:         groupInfo.Owner,
+		GroupName:     groupInfo.GroupName,
+		GroupId:       groupInfo.Id,
+		SourceType:    groupInfo.SourceType,
+		Members:       opts.Members,
+		Extra:         groupInfo.Extra,
```
Comment (guagualvcha) [inline x/storage/keeper/keeper.go:1501]: I think this `Extra` is not needed, as it is a field of a already exist group.

File: x/storage/keeper/keeper.go (hunk 9 above)
Comment (j75689) [reply x/storage/keeper/keeper.go:1271]: agreed, actually cross-chain package is not allow create group with members, so we can make them has the same behavior

File: proto/greenfield/storage/events.proto (hunk 7 above)
Comment (j75689) [reply proto/greenfield/storage/events.proto]: removed

File: x/storage/keeper/keeper.go (hunk 10 above)
Comment (j75689) [reply x/storage/keeper/keeper.go]: removed

File: x/permission/keeper/keeper.go (hunk 8 above)
Comment (j75689) [reply x/permission/keeper/keeper.go:308]: after the refactoring, the `ExpirationTime ` would not be `nil` it's only be set from user.

File: x/storage/keeper/cross_app_group.go [hunk 11]
```diff
@@ -374,6 +366,40 @@ func (app *GroupApp) handleUpdateGroupMemberSynPackage(ctx sdk.Context, header *
 	}
 }
 
+func (app *GroupApp) handleAddOrDeleteGroupMemberOperation(ctx sdk.Context, groupInfo *types.GroupInfo, updateGroupPackage *types.UpdateGroupMemberSynPackage) error {
+	options := types.UpdateGroupMemberOptions{
+		SourceType: types.SOURCE_TYPE_BSC_CROSS_CHAIN,
+	}
+	if updateGroupPackage.OperationType == types.OperationAddGroupMember {
+		options.MembersToAdd = updateGroupPackage.GetMembers()
+		options.MembersExpirationToAdd = updateGroupPackage.GetMemberExpiration()
```
Comment (guagualvcha) [inline x/storage/keeper/cross_app_group.go:375]: we need ensure the length of `MembersToAdd` and `MembersExpirationToAdd` are the same.

Comment (j75689) [reply x/storage/keeper/cross_app_group.go:375]: we had check function in `ValidateBasic` of `UpdateGroupMemberSynPackage`

https://github.com/j75689/greenfield/blob/3b2ed010df9480f1690591e19b3f510d3655d503/x/storage/types/crosschain.go#L968

File: x/permission/keeper/keeper.go [hunk 12]
```diff
@@ -283,14 +296,21 @@ func (k Keeper) VerifyPolicy(ctx sdk.Context, resourceID math.Uint, resourceType
...
+				if allowed && memberFound && groupMember.ExpirationTime != nil {
```
Comment (Fynnss) [inline x/permission/keeper/keeper.go]: Why not merge these lines to line 301?

File: app/upgrade.go [hunk 13]
```diff
@@ -36,3 +35,21 @@ func (app *App) RegisterUpgradeHandlers(chainID string, serverCfg *serverconfig.
 // 		},
 // 	)
 // }
+
+// registerNagquUpgradeHandler registers the upgrade handlers for Nagqu.
```
Comment (Fynnss) [inline app/upgrade.go:39]: please remove these useless code.

File: x/permission/keeper/keeper.go (hunk 12 above)
Comment (j75689) [reply x/permission/keeper/keeper.go]: fixed

File: app/upgrade.go (hunk 13 above)
Comment (j75689) [reply app/upgrade.go]: removed

**Review by guagualvcha — APPROVED** [review]

**Review by Fynnss — APPROVED** [review]
LGTM

**Pull request merged**